src/core/Hydrogen.cpp

```diff
@@ -2120,60 +2120,33 @@ void audioEngine_startAudioDrivers()
 
 
 	QString sAudioDriver = preferencesMng->m_sAudioDriver;
-	if ( sAudioDriver == "Auto" ) {
-	#ifndef WIN32
-		if ( ( m_pAudioDriver = createDriver( "JACK" ) ) == nullptr ) {
-			if ( ( m_pAudioDriver = createDriver( "ALSA" ) ) == nullptr ) {
-				if ( ( m_pAudioDriver = createDriver( "CoreAudio" ) ) == nullptr ) {
-					if ( ( m_pAudioDriver = createDriver( "PortAudio" ) ) == nullptr ) {
-						if ( ( m_pAudioDriver = createDriver( "OSS" ) ) == nullptr ) {
-							if ( ( m_pAudioDriver = createDriver( "PulseAudio" ) ) == nullptr ) {
-								audioEngine_raiseError( Hydrogen::ERROR_STARTING_DRIVER );
-								___ERRORLOG( "Error starting audio driver" );
-								___ERRORLOG( "Using the NULL output audio driver" );
-
-								// use the NULL output driver
-								m_pAudioDriver = new NullDriver( audioEngine_process );
-								m_pAudioDriver->init( 0 );
-							}
-						}
-					}
-				}
-			}
-		}
-	#else
-		//On Windows systems, use PortAudio is the prioritized backend
-		if ( ( m_pAudioDriver = createDriver( "PortAudio" ) ) == nullptr ) {
-			if ( ( m_pAudioDriver = createDriver( "ALSA" ) ) == nullptr ) {
-				if ( ( m_pAudioDriver = createDriver( "CoreAudio" ) ) == nullptr ) {
-					if ( ( m_pAudioDriver = createDriver( "JACK" ) ) == nullptr ) {
-						if ( ( m_pAudioDriver = createDriver( "OSS" ) ) == nullptr ) {
-							if ( ( m_pAudioDriver = createDriver( "PulseAudio" ) ) == nullptr ) {
-								audioEngine_raiseError( Hydrogen::ERROR_STARTING_DRIVER );
-								___ERRORLOG( "Error starting audio driver" );
-								___ERRORLOG( "Using the NULL output audio driver" );
-
-								// use the NULL output driver
-								m_pAudioDriver = new NullDriver( audioEngine_process );
-								m_pAudioDriver->init( 0 );
-							}
-						}
-					}
-				}
+	QStringList drivers = { "CoreAudio", "JACK", "ALSA", "OSS", "PulseAudio" };
+
+#ifdef WIN32
+	drivers.removeAll( "PortAudio" );
+	drivers.push_front( "PortAudio" );
+#endif
```

Comment:
I like the way you handle it with lists.

I wounder if it would be useful to make the choices a little more explicit by introducing another `#ifdef` for macOS and removing the drivers from the list, which are not supported by the OS anyway. At first glance it looks like Linux will use the CoreAudio driver or the second most common choice on Windows is ALSA. But on the other hand, with a little audio context these things are also somewhat evident.

Reply:
Thinking about it more, I think it's probably best overall for legibility and to make the intention clear to just provide the whole list for each platform separately. That should avoid confusion like this, makes the order easy to understand for each platform, and I think is probably actually fewer lines of code:
```c++
#if defined(WIN32)
    QStringList drivers = { "PortAudio", "JACK" };
#else if defined(__APPLE__)
    QStringList drivers = { "CoreAudio", "JACK", "PulseAudio" };
#else /* Linux */
    QStringList drivers = { "JACK", "ALSA", "OSS", "PulseAudio", "PortAudio" };
#endif
```